base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs

```diff
@@ -667,70 +704,76 @@ impl LMDBDatabase {
-                    lmdb_delete(&txn, &self.kernel_mmr_size_index, &v.kernel_mmr_size.to_be_bytes())?;
-                    lmdb_delete(&txn, &self.output_mmr_size_index, &v.output_mmr_size.to_be_bytes())?;
-                }
-            },
-            DbKey::BlockHash(_) => {
-                unimplemented!("Not supported. Use delete by height");
-            },
-            DbKey::OrphanBlock(k) => {
-                if let Some(orphan) = lmdb_get::<_, Block>(&txn, &self.orphans_db, &k)? {
-                    let parent_hash = orphan.header.prev_hash;
-                    lmdb_delete_key_value(&txn, &self.orphan_parent_map_index, parent_hash.as_slice(), &k)?;
-                    let tip: Option<Vec<u8>> = lmdb_get(&txn, &self.orphan_chain_tips_db, &k)?;
-                    if tip.is_some() {
-                        if lmdb_get::<_, Block>(&txn, &self.orphans_db, parent_hash.as_slice())?.is_some() {
-                            lmdb_insert(
-                                &txn,
-                                &self.orphan_chain_tips_db,
-                                parent_hash.as_slice(),
-                                &parent_hash,
-                                "orphan_chain_tips_db",
-                            )?;
-                        }
-                        lmdb_delete(&txn, &self.orphan_chain_tips_db, &k)?;
-                    }
-                    lmdb_delete(&txn, &self.orphans_db, k.as_slice())?;
+        // TODO: This can maybe be removed for performance if the check for block_accumulated_data existing is
+        // sufficient
+
+        let hash = header.hash();
+        // Check that there are no utxos or kernels linked to this.
+
+        if !lmdb_fetch_keys_starting_with::<TransactionKernelRowData>(hash.to_hex().as_str(), &txn, &self.kernels_db)?
+            .is_empty()
+        {
+            return Err(ChainStorageError::InvalidOperation(
+                "Cannot delete header because there are kernels linked to it".to_string(),
+            ));
+        }
+        if !lmdb_fetch_keys_starting_with::<TransactionOutputRowData>(hash.to_hex().as_str(), &txn, &self.utxos_db)?
+            .is_empty()
+        {
+            return Err(ChainStorageError::InvalidOperation(
+                "Cannot delete header because there are utxos linked to it".to_string(),
+            ));
+        }
+
+        lmdb_delete(&txn, &self.block_hashes_db, &hash)?;
+        lmdb_delete(&txn, &self.headers_db, &height)?;
+        lmdb_delete(&txn, &self.header_accumulated_data_db, &height)?;
+        lmdb_delete(&txn, &self.kernel_mmr_size_index, &header.kernel_mmr_size.to_be_bytes())?;
+        lmdb_delete(&txn, &self.output_mmr_size_index, &header.output_mmr_size.to_be_bytes())?;
+
+        Ok(())
+    }
+
+    fn delete_orphan(&mut self, txn: &WriteTransaction<'_>, hash: HashOutput) -> Result<(), ChainStorageError> {
+        if let Some(orphan) = lmdb_get::<_, Block>(&txn, &self.orphans_db, hash.as_slice())? {
```

Comment:
I started writing the check but then thought about it. There is another process that cleans up orphans and it would be difficult to synchronise these in code. The intention here is to prevent mangling of data by different threads resulting in an inconsistent state. I don't think that can happen in this case